### PR TITLE
dev-libs/libsodium: add multilib support

### PR DESCRIPTION
**Pull request by ConiKost**
dev-libs/libsodium: add multilib support

Closes: https://bugs.gentoo.org/622140

### CODE REVIEW

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 1]
```diff
@@ -1,9 +1,9 @@
-# Copyright 1999-2017 Gentoo Foundation
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: Firstly, please don't change the old versions and instead just add it to the newest.

Secondly, you've got wrong version in the commit message. Also, we don't add version to the prefix; instead, mention it in the commit message. However, if you only add it to the newest then there's no need for version at all.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 2]
```diff
@@ -16,12 +16,13 @@ IUSE="+asm minimal static-libs +urandom cpu_flags_x86_sse4_1 cpu_flags_x86_aes"
...
 
-src_prepare() {
+multilib_src_prepare() {
 	default
 	eautoreconf
 }
 
-src_configure() {
+multilib_src_configure() {
+	ECONF_SOURCE=${S}
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: `local`, please.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 3]
```diff
@@ -16,12 +16,13 @@ IUSE="+asm minimal static-libs +urandom cpu_flags_x86_sse4_1 cpu_flags_x86_aes"
...
 	local myconf
 
 	# --disable-pie needed on x86, bug #512734
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: `use x86` won't work correctly for multilib, compare against `${ABI}` instead.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 4]
```diff
@@ -37,7 +38,7 @@ src_configure() {
 		${myconf}
 }
 
-src_install() {
+src_multilib_install() {
 	default
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: `default` is not valid here, it includes a possible `make` call which is undesired. Call `einstalldocs` directly instead.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 5]
```diff
@@ -37,7 +38,7 @@ src_configure() {
...
 	prune_libtool_files
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: While at it, could you look if all `.la` files could be stripped and replace it with:

    find "${D}" -name '*.la' -delete || die

please?

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 2 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Fixed with newest commit.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 3 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Fixed with newest commit.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 4 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Okay, the "default" is from that example: https://wiki.gentoo.org/wiki/Project:Multilib/multilib-minimal

I've changed this now to emake DESTDIR="${D}" install, to make the deletion of *.la possible.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 5 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Well, when I do delete all *.la files, I am still able to compile unbound, so this should be ok?

**Comment by ConiKost** [comment]
@mgorny Thank you very much for you review. There is one point left, where I don't know, how to fix this.

There is already a cpuflags-patch included, to make use of cpu_flags_x86_sse4_1 and cpu_flags_x86_aes.
When I do emerge libsodium without multilib, it work's as expected.

[code]>>> Preparing source in /var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16 ...
 * Applying libsodium-1.0.10-cpuflags.patch ...                                                                                                                                                         [ ok ]
 * Running eautoreconf in '/var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16' ...
 * Running libtoolize --install --copy --force --automake ...                                                                                                                                           [ ok ]
 * Running aclocal -I m4 ...                                                                                                                                                                            [ ok ]
 * Running autoconf --force ...                                                                                                                                                                         [ ok ]
 * Running automake --add-missing --copy --foreign --force-missing ...                                                                                                                                  [ ok ]
 * Running elibtoolize in: libsodium-1.0.16/
 *   Applying portage/1.2.0 patch ...
 *   Applying sed/1.5.6 patch ...
 *   Applying as-needed/2.4.2 patch ...
 * Running elibtoolize in: libsodium-1.0.16/build-aux/
 *   Applying portage/1.2.0 patch ...
 *   Applying sed/1.5.6 patch ...
 *   Applying as-needed/2.4.3 patch ...
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16 ...
 * econf: updating libsodium-1.0.16/config.sub with /usr/share/gnuconfig/config.sub
 * econf: updating libsodium-1.0.16/config.guess with /usr/share/gnuconfig/config.guess
 * econf: updating libsodium-1.0.16/build-aux/config.sub with /usr/share/gnuconfig/config.sub
 * econf: updating libsodium-1.0.16/build-aux/config.guess with /usr/share/gnuconfig/config.guess
./configure --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --disable-dependency-tracking --disable-silent-rules --docdir=/usr/share/doc/libsodium-1.0.16 --htmldir=/usr/share/doc/libsodium-1.0.16/html --libdir=/usr/lib64 --enable-asm --disable-minimal --disable-blocking-random --disable-static --enable-sse4_1 --enable-aesni
configure: loading site script /usr/share/config.site[/code]

When running with multilib enabled, it seems not to recognize the new options, but the patch itself applies fine.

[code]>>> Preparing source in /var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16 ...
 * Applying libsodium-1.0.10-cpuflags.patch ...                                                                                                                                                         [ ok ]
>>> Source prepared.
>>> Configuring source in /var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16 ...
 * abi_x86_32.x86: running multilib-minimal_abi_src_configure
 * econf: updating libsodium-1.0.16/config.sub with /usr/share/gnuconfig/config.sub
 * econf: updating libsodium-1.0.16/config.guess with /usr/share/gnuconfig/config.guess
 * econf: updating libsodium-1.0.16/build-aux/config.sub with /usr/share/gnuconfig/config.sub
 * econf: updating libsodium-1.0.16/build-aux/config.guess with /usr/share/gnuconfig/config.guess
/var/tmp/portage/dev-libs/libsodium-1.0.16/work/libsodium-1.0.16/configure --prefix=/usr --build=i686-pc-linux-gnu --host=i686-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --disable-dependency-tracking --disable-silent-rules --docdir=/usr/share/doc/libsodium-1.0.16 --htmldir=/usr/share/doc/libsodium-1.0.16/html --libdir=/usr/lib --enable-asm --disable-minimal --disable-blocking-random --disable-static --enable-sse4_1 --enable-aesni --disable-pie
configure: WARNING: unrecognized options: --enable-sse4_1, --enable-aesni
configure: loading site script /usr/share/config.site[/code]

So multilib seems to do something different.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 4 above)
Comment (mgorny) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Oh, I'm sorry about this, I've misread it as `src_multilib_install_all`. The `default` was correct indeed.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 6]
```diff
@@ -16,16 +16,19 @@ IUSE="+asm minimal static-libs +urandom cpu_flags_x86_sse4_1 cpu_flags_x86_aes"
 
 PATCHES=( "${FILESDIR}"/${PN}-1.0.10-cpuflags.patch )
 
-src_prepare() {
+multilib_src_prepare() {
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: There's no multilib version of `src_prepare`, so revert this change and your problem should be gone ;-).

Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Whoops ;D Fixed in newest commit. Thanks!

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 4 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Okay, but should I revert it? When using "default", the find for deleting '*.la' seems not to run for me?

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 7]
```diff
@@ -1,9 +1,9 @@
-# Copyright 1999-2017 Gentoo Foundation
+# Copyright 1999-2018 Gentoo Foundation
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=6
 
-inherit eutils autotools
+inherit multilib-minimal eutils autotools
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: Please sort inherits lexically. Also, I think you don't use `eutils` (anymore).

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 8]
```diff
@@ -21,11 +21,14 @@ src_prepare() {
...
 }
 
-src_configure() {
+multilib_src_configure() {
+	local ECONF_SOURCE="${S}"
 	local myconf
 
 	# --disable-pie needed on x86, bug #512734
-	use x86 && myconf="${myconf} --disable-pie"
+	if [[ ${ABI} == "x86" ]]; then
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: I'm sorry but it just occurred to me that `${MULTILIB_ABI_FLAG}` would actually be better here as `${ABI}` differs e.g. between Linux and FreeBSD while the flag-var always will have `abi_x86_32`.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 9]
```diff
@@ -37,7 +40,10 @@ src_configure() {
 		${myconf}
 }
 
-src_install() {
-	default
-	prune_libtool_files
+multilib_src_install() {
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: You can remove `multilib_src_install` since the default eclass implementation will just work™.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 10]
```diff
@@ -37,7 +40,10 @@ src_configure() {
...
+	emake DESTDIR="${D}" install
+}
+
+multilib_src_install_all() {
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: You probably want `einstalldocs` here.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 4 above)
Comment (mgorny) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Ok, let me refocus here. Remove `multilib_src_install` = make it run the default, and do removal+docs in `_all`. Technically you could do it the other way around but then you'd unnecessarily run `find` twice.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 7 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Sure, will do. I thought, it was on purpose, as the original ebuild was reversed lexically ;-)

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 8 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Okay, will be changed. Should I always use MULTILIB_ABI_FLAG?

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 9 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: I will remove it.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 10 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: I've added it, works now locally for me.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 4 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Okay, will do :)

**Comment by ConiKost** [comment]
@mgorny: I've pushed now a new commit for libsodium.

> Also, please squash the commits into one afterwards.
As I've filled a couple of other PRs for other packages, should I always do all changes in one single commit, even with changes like adding a new license to portage?

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 8 above)
Comment (mgorny) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Inside `multilib_*` phases, yes. Outside that it's not defined ;-).

**Comment by mgorny** [comment]
> As I've filled a couple of other PRs for other packages, should I always do all changes in one single commit, even with changes like adding a new license to portage?

Generally, you should think how to form the commit history so that it will be helpful to others (and yourself in the future). [GLEP 66 s. 3.4](https://www.gentoo.org/glep/glep-0066.html#splitting-commits) gives some generic tips but it's up to you how to do it exactly.

Things like adding a new license are gray-ish. I would do it as a separate commit preceding the package but I wouldn't complain if you decided to squash it into the commit adding the package itself.

File: dev-libs/libsodium/libsodium-1.0.16.ebuild [hunk 11]
```diff
@@ -21,11 +21,14 @@ src_prepare() {
...
 
-src_configure() {
+multilib_src_configure() {
+	local ECONF_SOURCE="${S}"
 	local myconf
 
-	# --disable-pie needed on x86, bug #512734
-	use x86 && myconf="${myconf} --disable-pie"
+	# --disable-pie is needed on x86, see bug #512734
+	if [[ ${MULTILIB_ABI_FLAG} == "x86" ]]; then
```
Comment (mgorny) [inline dev-libs/libsodium/libsodium-1.0.16.ebuild]: That's not the correct value.

**Comment by ConiKost** [comment]
> Generally, you should think how to form the commit history so that it will be helpful to others (and yourself in the future). GLEP 66 s. 3.4 gives some generic tips but it's up to you how to do it exactly.
> 
> Things like adding a new license are gray-ish. I would do it as a separate commit preceding the package but I wouldn't complain if you decided to squash it into the commit adding the package itself.

Thanks! I see, what you mean. Thats helps me :)

File: dev-libs/libsodium/libsodium-1.0.16.ebuild (hunk 11 above)
Comment (ConiKost) [reply dev-libs/libsodium/libsodium-1.0.16.ebuild]: Fixed and pushed -r1.
